heimlig/src/hsm/keystore.rs

```diff
@@ -106,50 +134,117 @@ impl<const STORAGE_SIZE: usize, const NUM_KEYS: usize> KeyStore
+            Some(key_layout) => {
+                if key_layout.actual_size == 0 {
                     return Err(Error::KeyNotFound);
                 }
-                Ok(key_info.actual_size)
+                Ok(key_layout.actual_size)
             }
         }
     }
 
     fn import(&mut self, id: KeyId, src: &[u8]) -> Result<(), Error> {
-        match self.infos.iter_mut().find(|key_info| key_info.id == id) {
+        match self
+            .layout
+            .iter_mut()
+            .find(|key_layout| key_layout.info.id == id)
+        {
             None => Err(Error::InvalidKeyId),
-            Some(key_info) => {
-                if src.len() > key_info.max_size {
+            Some(key_layout) => {
+                if src.len() > key_layout.info.max_size {
                     return Err(Error::BufferTooLarge);
                 }
-                let dest = &mut self.storage[key_info.offset..(key_info.offset + src.len())];
+                if !key_layout.info.permissions.import {
+                    return Err(Error::NotAllowed);
+                }
+                let offset = key_layout.offset;
+                let size = src.len();
+                let dest = &mut self.storage[offset..(offset + size)];
                 dest.copy_from_slice(src);
-                key_info.actual_size = src.len();
+                key_layout.actual_size = src.len();
                 Ok(())
             }
         }
     }
 
     fn export<'data>(&self, id: KeyId, dest: &'data mut [u8]) -> Result<&'data [u8], Error> {
-        match self.infos.iter().find(|key_info| key_info.id == id) {
+        match self
+            .layout
+            .iter()
+            .find(|key_layout| key_layout.info.id == id)
+        {
+            None => Err(Error::InvalidKeyId),
+            Some(key_layout) => {
+                if !key_layout.info.permissions.export {
+                    return Err(Error::NotAllowed);
+                }
+                self.export_unchecked(id, dest)
+            }
+        }
+    }
+
+    fn delete(&mut self, id: KeyId) -> Result<(), Error> {
+        match self
+            .layout
+            .iter_mut()
+            .find(|key_layout| key_layout.info.id == id)
+        {
             None => Err(Error::InvalidKeyId),
-            Some(key_info) => {
-                if key_info.actual_size == 0 {
+            Some(key_layout) => {
+                if !key_layout.info.permissions.delete {
+                    return Err(Error::NotAllowed);
+                }
+                if key_layout.actual_size == 0 {
                     return Err(Error::KeyNotFound);
                 }
-                if dest.len() < key_info.actual_size {
+                let offset = key_layout.offset;
+                let size = key_layout.actual_size;
+                let key = &mut self.storage[offset..(offset + size)];
+                key.fill(0);
+                key_layout.actual_size = 0;
```

Comment:
After the key is deleted. should we mark it somewhere as such?

Reply:
We currently use `key_layout.actual_size == 0` to determine this (see `is_stored`). I feel like an additional status flag would just duplicate state with `actual_size`.